### PR TITLE
Use column names instead of field refs for pivot viz settings

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/49405

There are 2 last viz settings that use field refs instead of column names:
```
{
  "pivot_table.column_split": {
    "rows": [["field", 1, null]],
    "columns": [["field", 2, null]],
    "values": [["aggregation", 0]],
  },
  "pivot_table.collapsed_rows": {
    "rows": [["field", 1, null]],
    "values": ["Gadget"],
  }
}
```

They cause the following issues:
- Applying a temporal unit parameter resets the viz settings since the ref changes (and fixing this is the reason for doing this work)
- When the QP switches between string and integer refs all pivot viz settings are lost
- Any change to field refs require us to write a custom migration or migrate data on read. Since we're reworking field refs now this can become a blocker.
- We cannot sync field ref changes when the query is updated (but we can with column names, e.g. `ID_2` -> `ID_3` changes)

This PR makes these settings work as others, where column names are used, which are synced with the query. Key parts of the solution:
- There is no migration. We support both old- and new-style viz settings both on the FE and the BE. It's not possible to write a custom migration for this since `result_metadata` is broken for pivot queries. It doesn't contain metadata for some of the returned columns, usually aggregations, but not only them.
- The FE uses new column name-based settings for new pivot questions. The FE doesn't change the settings for existing questions until they are explicitly modified, e.g. the columns are moved between rows/columns/values.
- Serialization supports both old and new-style settings. Column name settings are saved as is, while legacy settings are transformed according to the existing logic.
- Most tests were updated to use column names for viz settings to make sure we have enough coverage. There are still tests for the happy path and edge cases with field refs to make sure that legacy questions are supported.
- It's still an open question if we need to run a db migration or if that's already good enough. It's not possible to add a "custom migration" but we could use `quartz` in the future to rewrite cards in the background.

New settings look like this:
```
{
  "pivot_table.column_split": {
    "rows": ["CATEGORY"],
    "columns": ["VENDOR"],
    "values": ["sum"],
  },
  "pivot_table.collapsed_rows": {
    "rows": ["CATEGORY"],
    "values": ["Gadget"],
  }
}
```

How to verify:
- Checkout `master` and create pivot questions with several aggregations & breakouts. Make sure to test both default & custom column split (i.e. drag columns between sections in the viz settings)
- Checkout this branch and open your saved questions. There should be no changes. Your questions should not become ad-hoc.
- Modify pivot column split for some of the existing questions. Make sure that the settings have been migrated to column names and that it works as expected.
- Create new pivot questions. Make sure that column names are used by default.

### CODE REVIEW

**Comment by bshepherdson** [comment]
> we could use quartz in the future to rewrite cards in the background

Hopefully that won't be needed for this specific change. We have been talking more broadly about changes to how queries relate to visualizations, and it seems likely to me that the current system for viz settings will be completely replaced in the medium term.

File: frontend/src/metabase/visualizations/visualizations/PivotTable/PivotTable.unit.spec.tsx [hunk 1]
```diff
@@ -124,13 +124,13 @@ describe("Visualizations > PivotTable > PivotTable", () => {
       });
 
       it("expanding collapsed columns", async () => {
-        const hiddenSettings = {
+        const hiddenSettings: VisualizationSettings = {
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/visualizations/PivotTable/PivotTable.unit.spec.tsx:127]: `createMockVisualizationSettings`?

File: frontend/src/metabase/visualizations/visualizations/PivotTable/PivotTable.tsx [hunk 2]
```diff
@@ -281,14 +281,18 @@ function _PivotTable({
   };
 
   const leftHeaderWidth =
-    pivoted?.rowIndexes.length > 0
+    (pivoted?.rowIndexes?.length ?? 0) > 0
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/visualizations/PivotTable/PivotTable.tsx:284]: prev version should work as well, but your is more explicit

File: frontend/src/metabase/visualizations/components/settings/ChartSettingFieldsPartition.unit.spec.js [hunk 3]
```diff
@@ -26,10 +26,14 @@ describe("ChartSettingFieldsPartition", () => {
           { name: "rows", title: "Rows", columnFilter: x => Boolean(x) },
         ]}
         columns={[
-          createMockColumn({ field_ref: fieldRef, display_name: fieldName }),
+          createMockColumn({
+            name: fieldName,
+            field_ref: fieldRef,
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/settings/ChartSettingFieldsPartition.unit.spec.js:31]: can we drop `field_ref` here?

File: src/metabase/query_processor/pivot.clj [hunk 4]
```diff
@@ -347,9 +347,32 @@
...
-  settings and generates pivot-rows and pivot-cols to use for generating subqueries."
+(mu/defn- column-name-pivot-options :- ::pivot-opts
+  "Looks at the `pivot_table.column_split` key in the card's visualization settings and generates `pivot-rows` and
+  `pivot-cols` to use for generating subqueries. Supports column name-based settings only."
+  [query        :- [:map
+                    [:database ::lib.schema.id/database]]
+   viz-settings :- [:maybe :map]]
+  (let [{:keys [rows columns values]} (:pivot_table.column_split viz-settings)
+        metadata-provider  (or (:lib/metadata query)
+                               (lib.metadata.jvm/application-database-metadata-provider (:database query)))
```
Comment (bshepherdson) [inline src/metabase/query_processor/pivot.clj:358]:
```suggestion
        metadata-provider  (or (:lib/metadata query)
                               (qp.store/metadata-provider)
                               (lib.metadata.jvm/application-database-metadata-provider (:database query)))
```
in the QP there's usually a dynamically bound metadata provider we can reuse.

File: frontend/src/metabase/visualizations/visualizations/PivotTable/utils.ts [hunk 5]
```diff
@@ -26,58 +26,71 @@ import {
...
     for (const { columnFilter: filter, name } of partitions) {
-      const column = columns.find(
-        c => JSON.stringify(c.field_ref) === fieldRef,
-      );
-      if (filter == null || filter(column)) {
-        value[name].push(column?.field_ref as FieldReference);
+      const column = columns.find(c => c.name === columnName);
+      if (column != null && (filter == null || filter(column))) {
+        value[name] = value[name] ?? [];
+        value[name]?.push(column.name);
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/visualizations/PivotTable/utils.ts:70]:
```suggestion
        value[name].push(column.name);
```

**Review by uladzimirdev — APPROVED** [review]
FE part LGTM

File: src/metabase/query_processor/pivot.clj [hunk 6]
```diff
@@ -387,6 +410,22 @@
...
+
+  Field ref-based visualization settings are considered legacy and are not used for new questions. To not break existing
+  questions we need to support both old- and new-style settings until they are fully migrated."
+  [query        :- [:map
+                    [:database ::lib.schema.id/database]]
+   viz-settings :- [:maybe :map]]
+  (let [column-split         (:pivot_table.column_split viz-settings)
+        column-split-rows    (seq (:rows column-split))
+        column-split-columns (seq (:columns column-split))]
+    (if (and (every? string? column-split-rows) (every? string? column-split-columns))
```
Comment (bshepherdson) [inline src/metabase/query_processor/pivot.clj:425]:
```suggestion
  (let [{:keys [rows columns]} (:pivot_table.column_split viz-settings)]
    (if (and (every? string? rows) (every? string? columns))
```
You shouldn't need to explicitly call `seq` on the pieces (even if they're maybe `nil`), the implementation of `every?` will do it if needed.

**Review by bshepherdson — APPROVED** [review]

File: frontend/src/metabase/visualizations/visualizations/PivotTable/utils.ts (hunk 5 above)
Comment (ranquild) [reply frontend/src/metabase/visualizations/visualizations/PivotTable/utils.ts]: Fixed

File: frontend/src/metabase/visualizations/components/settings/ChartSettingFieldsPartition.unit.spec.js (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/visualizations/components/settings/ChartSettingFieldsPartition.unit.spec.js]: Yes, fixed

File: frontend/src/metabase/visualizations/visualizations/PivotTable/PivotTable.unit.spec.tsx (hunk 1 above)
Comment (ranquild) [reply frontend/src/metabase/visualizations/visualizations/PivotTable/PivotTable.unit.spec.tsx]: Fixed

**Review by sansolovyov — APPROVED** [review]
this is fantastic

**Review by adam-james-v — APPROVED** [review]

**Pull request merged**